src/treelearner/gpu_tree_learner.cpp

```diff
@@ -0,0 +1,1064 @@
+  data_size_t num_data_in_right_child = GetGlobalDataCountInLeaf(right_leaf);
+  // only have root
+  if (right_leaf < 0) {
+    smaller_leaf = -1;
+  } else if (num_data_in_left_child < num_data_in_right_child) {
+    smaller_leaf = left_leaf;
+  } else {
+    smaller_leaf = right_leaf;
+  }
+
+  // Copy indices, gradients and hessians as early as possible
+  if (smaller_leaf >= 0) {
+    // only need to initialize for smaller leaf
+    // Get leaf boundary
+    const data_size_t* indices = data_partition_->indices();
+    data_size_t begin = data_partition_->leaf_begin(smaller_leaf);
+    data_size_t end = begin + data_partition_->leaf_count(smaller_leaf);
+
+    // copy indices to the GPU:
+    #if GPU_DEBUG >= 2
+    Log::Info("Copying indices, gradients and hessians to GPU...");
+    printf("indices size %d being copied (left = %d, right = %d)\n", end - begin,num_data_in_left_child,num_data_in_right_child);
+    #endif
+    indices_future_ = boost::compute::copy_async(indices + begin, indices + end, device_data_indices_->begin(), queue_);
+
+    #pragma omp parallel for schedule(static)
+    for (data_size_t i = begin; i < end; ++i) {
+      ordered_hessians_[i - begin] = hessians_[indices[i]];
+    }
+    // copy ordered hessians to the GPU:
+    hessians_future_ = queue_.enqueue_write_buffer_async(device_hessians_, 0, (end - begin) * sizeof(score_t), ptr_pinned_hessians_);
+
+    #pragma omp parallel for schedule(static)
+    for (data_size_t i = begin; i < end; ++i) {
+      ordered_gradients_[i - begin] = gradients_[indices[i]];
+    }
+    // copy ordered gradients to the GPU:
+    gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, (end - begin) * sizeof(score_t), ptr_pinned_gradients_);
+
+    #if GPU_DEBUG >= 2
+    Log::Info("gradients/hessians/indiex copied to device with size %d", end - begin);
+    #endif
+  }
+  return SerialTreeLearner::BeforeFindBestSplit(tree, left_leaf, right_leaf);
+}
+
+bool GPUTreeLearner::ConstructGPUHistogramsAsync(
+  const std::vector<int8_t>& is_feature_used,
+  const data_size_t* data_indices, data_size_t num_data,
+  const score_t* gradients, const score_t* hessians,
+  score_t* ordered_gradients, score_t* ordered_hessians) {
+
+  if (num_data <= 0) {
+    return false;
+  }
+
+  
+  // copy data indices if it is not null
+  if (data_indices != nullptr && num_data != num_data_) {
+    indices_future_ = boost::compute::copy_async(data_indices, data_indices + num_data, device_data_indices_->begin(), queue_);
+  }
+  // generate and copy ordered_gradients if gradients is not null
+  if (gradients != nullptr) {
+    if (num_data != num_data_) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data; ++i) {
+        ordered_gradients[i] = gradients[data_indices[i]];
+      }
+      gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, num_data * sizeof(score_t), ptr_pinned_gradients_);
```

Comment:
Because I want to start data copying process as early as possible, hopefully overlap the copying process with other task on CPU and hide the data transfer cost. Currently, in most cases the copying procedure entirely overlaps with `SerialTreeLearner::tBeforeFindBestSplit`. We only wait for these event futures (`indices_future_`, `gradients_future_` and `hessians_future_`) immediately before launching the GPU kernel, and hopefully at that time all data has been transferred to GPU and the overhead for data transfer is close to 0.